[DOCS] Update ES|QL generated docs to consistently use the `applies_to` metadata

craigtaverner opened the pull request:
## Summary

This PR updates the documentation generation system for ESQL functions to correctly generate [`applies_to` metadata](https://elastic.github.io/docs-builder/syntax/applies/) in DocsV3, namely to support multiple versions and lifecycles.

This metadata is needed to signal version differentiation in the new docs, because we don't publish version-branched docs in 9.x (we don't duplicate every page for every minor). 

For example, if something is preview in 9.0 and GA in 9.1 we need all this information on single, live 9.x page, that's what `applies_to` enables. Importantly, this also enables us to correctly signal serverless feature availability in the new docs.

We need different levels of `applies_to` discrimination for this:
- [page annotations](https://elastic.github.io/docs-builder/syntax/applies/#page-annotations)
- [section annotations](https://elastic.github.io/docs-builder/syntax/applies/#section-annotations)
-  [inline annotations](https://elastic.github.io/docs-builder/syntax/applies/#inline-annotations).

> [!IMPORTANT]
> Once this PR is merged, we will use this new documentation generation logic to properly GA full-text search functions in a [follow-up PR.](https://github.com/elastic/elasticsearch/pull/129786)

## Key changes

### 1. Documentation generation code

- `x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/Example.java` - Added `applies_to()` field for version-specific examples
- `x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3Support.java` - Updated to generate standardized `applies_to` blocks instead of manual callouts

> [!NOTE]
> `DocsV3Support.java` contains the most significant docs gen code changes:
> - **Total additions**: ~57 lines
> - **Total deletions**: ~63 lines  
> - **Total line edits**: ~120 lines
> - **Net change**: ~-6 lines (the file got slightly smaller)


### 2. Function status corrections

- **All full-text search functions**: now correctly marked as preview with proper `applies_to` syntax
- **Spatial functions** (`ST_*`): Added missing `preview = true` flags, and now use  proper `applies_to` syntax 
- A few other functions weren't correctly identifying inline version changes such as ToUpper supporting MVs in 9.1 only

### 3. Generated documentation updates

As a result all of the function code adjusted results in updated markdown and Kibana JSON file updates

## Result

The documentation system now correctly generates version-aware metadata, enabling proper lifecycle tracking across versions. All full-text search functions are consistently marked as technical preview with the correct `applies_to` syntax, to enable us flipping them to GA for 9.1.


# 👀  [URL preview](https://docs-v3-preview.elastic.dev/elastic/elasticsearch/pull/128576/reference/query-languages/esql/esql-functions-operators)

craigtaverner committed:
WIP try make more use of real applies_to

Commit:
[CI] Auto commit changes from spotless

leemthompo committed:
Add @FunctionAppliesTo annotations to functions for new docs system

- Add PREVIEW annotations to all preview functions
- Update docs generation logic to use annotations instead of preview boolean
- Keep existing preview boolean for Kibana

leemthompo committed:
Tweak multimatch applies_to

leemthompo committed:
Manually update list

Commit:
[CI] Auto commit changes from spotless

leemthompo committed:
Changed stack: ga 9.1 to stack: coming in multiple places

Match.java: Updated the options parameter description
MultiMatch.java: Updated the options parameter description
ToLower.java: Reformatted parameter description and updated version annotation
ToUpper.java: Removed the appliesTo annotation entirely and reformatted parameter description

leemthompo committed:
Merge branch 'esql_docs_real_applies_to' of https://github.com/craigtaverner/elasticsearch into esql_docs_real_applies_to

leemthompo committed:
fix conflict markers

leemthompo committed:
fix tests

Commit:
[CI] Auto commit changes from spotless

leemthompo committed:
remove inline applies_tos for time being

Commit:
[CI] Auto commit changes from spotless

leemthompo committed:
mismatch between local generation and CI generation?

leemthompo committed:
Revert "mismatch between local generation and CI generation?"

This reverts commit 183b1318cfbbd758ff5d205edac96578739901de.

leemthompo committed:
Add generated files, cleanup errors

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

leemthompo committed:
Update lifecycles, cleanup

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MultiMatch.java:
```diff
@@ -137,18 +137,12 @@ public class MultiMatch extends FullTextFunction implements OptionalArgument, Po
-            Use `MULTI_MATCH` to perform a <<query-dsl-multi-match-query,multi-match query>> on the specified field.
-            The multi_match query builds on the match query to allow multi-field queries.""",
+        appliesTo = { @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.COMING) },
+        description = "Use `MULTI_MATCH` to perform a <<query-dsl-multi-match-query,multi-match query>> on the specified field."
```
I think I had to break up into separate strings because of check style violation but definitely need to fix the incorrect formatting

leemthompo replied:
https://github.com/elastic/elasticsearch/pull/128576/commits/296cae6c70a07edf506dc3010b4bcdf190fa7d38

leemthompo committed:
fix spacing and line formatting in the  function

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

leemthompo commented on docs/reference/query-languages/esql/kibana/definition/functions/multi_match.json:
```diff
@@ -23,7 +23,7 @@
           "mapParams" : "{name='fuzziness', values=[AUTO, 1, 2], description='Maximum edit distance allowed for matching.'}, {name='auto_generate_synonyms_phrase_query', values=[true, false], description='If true, match phrase queries are automatically created for multi-term synonyms. Defaults to true.'}, {name='minimum_should_match', values=[2], description='Minimum number of clauses that must m ...
           "optional" : true,
-          "description" : "(Optional) Additional options for MultiMatch, passed as <<esql-function-named-params,function named parameters>>.\"  See <<query-dsl-multi-match-query,multi-match query>> for more information."
+          "description" : "(Optional) Additional options for MultiMatch, passed as <<esql-function-named-params,function named parameters>>.{applies_to}`stack: coming`\"  See <<query-dsl-multi-match-query,multi-match query>> for more information."
```
@drewdaemon @stratoula FYI we'll start having more of these little applies_to syntaxes in the ESQL docs, but y'all will need to filter them out for Kibana 

This example uses [inline](https://elastic.github.io/docs-builder/syntax/applies/#inline-applies-to) annotation: 

```markdown
{applies_to}`foo: bar`
```

There are also [section](https://elastic.github.io/docs-builder/syntax/applies/#sections) annotations: 

``````markdown
```{applies_to}
stack: ga 9.1
```
``````

stratoula replied:
@craigtaverner can you filter them out for us? Or replace with the correct value?

stratoula replied:
yes if ES can help as they do with other similar things would be highly appreciated

leemthompo replied:
@stratoula I think it's now stripping the applies_to metadata correctly from the Kibana files, can you have a look please?

Sorry about the noise from the "Do not edit this" change :/

leemthompo committed:
Promotes ESQL search functions from preview to GA in v9.1.0 and adds multi-valued string support, cleanup
• upgraded several esql search functions (match, kql, qstr) from preview to ga in version 9.1.0
• updated applies_to annotations to specify both preview 9.0.0 and ga 9.1.0 lifecycle stages
• added version-specific documentation examples with applies_to markers for ga 9.1.0 features
• enhanced to_lower and to_upper functions to support multi-valued expressions in ga 9.1.0
• updated multi_match function from "coming" to preview 9.0.0 status
• removed preview markers from search function lists and json definitions
• fixed typo in json generation comment ("do no edit" → "do not edit")
• added version guards around function parameters and descriptions using applies_to syntax
• updated function parameter descriptions to indicate ga 9.1.0 availability for named parameters

leemthompo committed:
Merge branch 'esql_docs_real_applies_to' of https://github.com/craigtaverner/elasticsearch into esql_docs_real_applies_to

leemthompo committed:
use detailedDescription + and fix match_phrase applies_to syntax

leemthompo committed:
strip applies to from kibana docs

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

carlosdelest commented on docs/reference/query-languages/esql/_snippets/functions/description/match.md:
```diff
@@ -2,5 +2,17 @@
+When Match is used on a [semantic_text](/reference/elasticsearch/mapping-reference/semantic-text.md) field, it will perform a semantic query on the field.
+
+Match can use [function named parameters](/reference/query-languages/esql/esql-syntax.md#esql-function-named-params) to specify additional options
+for the match query. {applies_to}`stack: ga 9.1.0`
```
> {applies_to}`stack: ga 9.1.0`

Is that necessary for the docs? Or in that location? It sounds a bit weird in terms of placement...

carlosdelest commented on docs/reference/query-languages/esql/_snippets/functions/examples/match.md:
````diff
@@ -15,6 +15,10 @@ FROM books
 
+```{applies_to}
+stack: ga 9.1.0
+```
````
Does it make sense to add tech preview for 8.17+ / 9.0? Same applies for the other functions

leemthompo replied:
can do it for 9.0 but we don't cover 8.x here

carlosdelest commented on docs/reference/query-languages/esql/_snippets/functions/parameters/match.md:
```diff
@@ -9,5 +9,5 @@
 
 `options`
-:   (Optional) Match additional options as [function named parameters](/reference/query-languages/esql/esql-syntax.md#esql-function-named-params). See [match query](/reference/query-languages/query-dsl/query-dsl-match-query.md) for more information.
+:   (Optional) Match additional options as [function named parameters](/reference/query-languages/esql/esql-syntax.md#esql-function-named-params). {applies_to}`stack: ga 9.1.0`
```
```suggestion
:   (Optional) Match additional options as [function named parameters](/reference/query-languages/esql/esql-syntax.md#esql-function-named-params). {applies_to}`stack: ga 9.1.0, preview 9.0.0`
```

leemthompo committed:
update roundto, matchphrase lifecycles

leemthompo committed:
fix match named params info

leemthompo committed:
fix qstr

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

stratoula reviewed (approved):
Kibana docs changes look great (I didnt check all of them, only the full text functions and the more recent ones)

carlosdelest reviewed (approved):
From the full text functions, LGTM - only a minor correction

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/Kql.java:
```diff
@@ -35,7 +37,9 @@ public class Kql extends FullTextFunction {
         returnType = "boolean",
-        preview = true,
+        appliesTo = {
+            @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.PREVIEW, version = "9.0.0"),
+            @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.GA, version = "9.1.0") },
```
Why does this come with 2 markers? Isn't the `GA in 9.1` information sufficient?

Also, what about `8.19`? I think most that's going live in 9.1 is also going live in 8.19, no?

alex-spies reviewed (changes requested):
I'm finding this to be really hard to review.

If I'm not mistaken, this contains
1. changes to the generation of the `preview`, `GA in 9.1` etc. markers
2. some editorial changes to function descriptions, some minor reformats etc.
3. changes to the preview status of some functions, notably marking full text functions as GA

To be able to review this better, and later be able to trace back our steps via git blame, I think it's important this PR gets split so it does fewer things.

Ideally, only 1., but 1. and 2. would also be okay. 3. really should go into its own PR, esp. since the PR's title and description don't indicate that this is where we mark full text functions as GA.

leemthompo commented on docs/reference/query-languages/esql/_snippets/lists/search-functions.md:
```diff
@@ -1,5 +1,5 @@
-* [preview] [`MATCH`](../../functions-operators/search-functions.md#esql-match)
-* [preview] [`MATCH_PHRASE`](../../functions-operators/search-functions.md#esql-match_phrase)
-* [preview] [`QSTR`](../../functions-operators/search-functions.md#esql-qstr)
+* [`KQL`](../../functions-operators/search-functions.md#esql-kql)
+* [`MATCH`](../../functions-operators/search-functions.md#esql-match)
+* [`MATCH_PHRASE`](../../functions-operators/search-functions.md#esql-match_phrase)
+* [`QSTR`](../../functions-operators/search-functions.md#esql-qstr)
```
> If I'm not mistaken, this contains

1. changes to the generation of the preview, GA in 9.1 etc. markers
2. some editorial changes to function descriptions, some minor reformats etc.
3. changes to the preview status of some functions, notably marking full text functions as GA

correct, unfortunately 3 is dependent on 1 because the applies_to markers are what makes it possible to distinguish between preview/GA features in the new docs system

we can't just remove preview banners, because the docs are versionless

we also needed 2. because the way these tags interact with the rest of the docs requires some changes, but there is a lot of noise generated as a side effect and very sorry about that

leemthompo replied:
I've now removed 3 from the scope of this PR

leemthompo replied:
and @nik9000 helped remove some of the Kibana noise from this PR due to the irrelevant typo fix 🙏

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Values.java:
```diff
@@ -72,6 +74,7 @@ public class Values extends AggregateFunction implements ToAggregator {
             "long",
             "version" },
         preview = true,
```
we need the boolean for Kibana and serverless flags

leemthompo replied:
in future once we're into the swing of using this metadata, the only thing a dev will need to do when something flips to GA is remove that boolean and generate the docs to remove the preview flag from the serverless pill 

you will already have set the appropriate applies_to metadata when first drafting the docs

and the docs build will take care of updating the stack versioned information

nik9000 replied:
We could fix this, I think. But the PR is already pretty big.

leemthompo replied:
it's handy for Kibana and serverless tbh

leemthompo commented on docs/reference/query-languages/esql/_snippets/functions/parameters/match.md:
```diff
@@ -9,5 +9,5 @@
 :   Value to find in the provided field.
 
 `options`
-:   (Optional) Match additional options as [function named parameters](/reference/query-languages/esql/esql-syntax.md#esql-function-named-params). See [match query](/reference/query-languages/query-dsl/query-dsl-match-query.md) for more information.
```
@ioanatia this link was removed here but I can't remember if I did it on purpose or not. I do remember you updated to add the list of function named parameters directly in the ES|QL match docs...

So maybe linking to https://www.elastic.co/docs/reference/query-languages/query-dsl/query-dsl-match-query is no longer needed?

ioanatia replied:
We did this because we are already listing all the match function options on the ES|QL docs.
The other reason is because the match function will diverge from the match query.
For example we are now looking at adding support in the match function to receive multiple fields.

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

leemthompo commented on docs/reference/query-languages/esql/_snippets/functions/parameters/to_upper.md:
```diff
@@ -3,5 +3,5 @@
 
 `str`
-:   String expression. If `null`, the function returns `null`. The input can be a single- or multi-valued column or an expression.
+:   String expression. If `null`, the function returns `null`. The input can be a single-valued column or expression, or a multi-valued column or expression {applies_to}`stack: ga 9.1.0`. 
```
Definitely take that as a follow up 👍

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

leemthompo committed:
spatial functions are preview

leemthompo committed:
unsigned long is preview

leemthompo committed:
revert GAing full-text functions, to come in followup

leemthompo committed:
update qstr

leemthompo committed:
multi-match docs arent published, reduce noise by reverting changes

nik9000 committed:
Merge branch 'main' into esql_docs_real_applies_to

nik9000 committed:
Merge remote-tracking branch 'craigtaverner/esql_docs_real_applies_to' into esql_docs_real_applies_to

leemthompo committed:
update manual list of search functions

leemthompo committed:
fix typo

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

leemthompo committed:
The ST_DISTANCE function is not tech-preview

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/ToUpper.java:
```diff
@@ -30,17 +28,12 @@ public class ToUpper extends ChangeCase {
         returnType = { "keyword" },
         description = "Returns a new string representing the input string converted to upper case.",
-        examples = @Example(file = "string", tag = "to_upper"),
-        appliesTo = {
-            @FunctionAppliesTo(
-                lifeCycle = FunctionAppliesToLifecycle.COMING,
-                version = "9.1.0",
-                description = "Support for multivalued parameters is only available from 9.1.0"
-            ) }
```
Shouldn't it say serverless, so we know it is not in stateful?

craigtaverner commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3SupportTests.java:
````diff
@@ -311,22 +311,9 @@ public void testRenderingLayoutFromClass() throws Exception {
-            ###### DISCONTINUED 9.0.0
-            The ability to count the number of emojis in a string has been discontinued since 9.0.0
-            :::
+            ```{applies_to}
````
This test originally deliberately tried to test a complex example, and the new version is extremely simple. Perhaps we could think of a new complex example? At least where stack and serverless are both mentioned?

leemthompo replied:
💯  in a follow up 👍

leemthompo replied:
💯 — will follow up

craigtaverner commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3SupportTests.java:
```diff
@@ -418,7 +404,7 @@ public static class TestClass extends Function {
                     To count the number of times an expression returns `TRUE` use a
                     <<esql-where>> command
-                    to remove rows that shouldn’t be included""", file = "stats", tag = "count-where"),
+                    to remove rows that shouldn't be included""", file = "stats", tag = "count-where"),
```
This is wrong. We took the decision in March to put in the correct typesetting apostrophe because MD does not render it correctly, while asciidoc does. This appears to reverse that decision.

leemthompo replied:
unintended side effect, and unfortunately a lot less important than being able to GA the search functions for 9.1

craigtaverner reviewed (commented):
Overall I think this is heading in the right direction and is an improvement, but I did see a number of preview items that seemed to be wrong. Also, I'm finding the use of preview confusing, sometimes with a version sometimes without, and the choice of version does not match any pattern I can discern.

leemthompo committed:
StX and StY are not preview

leemthompo commented on docs/reference/query-languages/esql/_snippets/functions/layout/st_distance.md:
````diff
@@ -1,6 +1,10 @@
 % This is generated by ESQL's AbstractFunctionTestCase. Do not edit it. See ../README.md for how to regenerate it.
 
 ## `ST_DISTANCE` [esql-st_distance]
+```{applies_to}
````
fixed — this was human error

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/Match.java:
```diff
@@ -136,28 +136,26 @@ public class Match extends FullTextFunction implements OptionalArgument, PostAna
-            Using `MATCH` is equivalent to using the `match` query in the Elasticsearch Query DSL.
-
+            Using `MATCH` is equivalent to using the `match` query in the Elasticsearch Query DSL.""",
+        detailedDescription = """
```
the markdown docs are all squished together as a result, if we choose to not include detailedDescription in the Kibana docs then yes this is downstream of that

leemthompo replied:
Will add to list for follow up

leemthompo commented on docs/reference/query-languages/esql/kibana/definition/functions/match.json:
```diff
@@ -2,7 +2,7 @@
   "type" : "scalar",
   "name" : "match",
-  "description" : "Use `MATCH` to perform a match query on the specified field.\nUsing `MATCH` is equivalent to using the `match` query in the Elasticsearch Query DSL.\n\nMatch can be used on fields from the text family like text and semantic_text,\nas well as other field types like keyword, boolean, dates, and numeric types.\nWhen Match is used on a semantic_text field, it will perform a semanti ...
+  "description" : "Use `MATCH` to perform a match query on the specified field.\nUsing `MATCH` is equivalent to using the `match` query in the Elasticsearch Query DSL.",
```
they belong in detailedDescription

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchOperator.java:
```diff
@@ -29,6 +31,9 @@ public class MatchOperator extends Match {
         preview = true,
+        appliesTo = {
+            // @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.PREVIEW, version = "9.0.0"),
+            @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.GA, version = "9.1.0") },
```
wait we don't generated operator docs so this logic is futile currently

leemthompo committed:
Revert incorrect preview/GA status changes

- Revert CATEGORIZE back to GA (remove preview flags that were incorrectly added)
- Revert spatial functions (ST_CONTAINS, ST_DISJOINT, ST_INTERSECTS, ST_WITHIN) back to GA
- Revert match operator to show PREVIEW status correctly
- Revert MatchOperator appliesTo annotation to PREVIEW lifecycle
- Update TO_LOWER/TO_UPPER parameter descriptions for clarity

leemthompo commented on docs/reference/query-languages/esql/kibana/definition/functions/st_contains.json:
```diff
@@ -152,6 +152,6 @@
     "FROM airport_city_boundaries\n| WHERE ST_CONTAINS(city_boundary, TO_GEOSHAPE(\"POLYGON((109.35 18.3, 109.45 18.3, 109.45 18.4, 109.35 18.4, 109.35 18.3))\"))\n| KEEP abbrev, airport, region, city, city_location"
   ],
-  "preview" : false,
+  "preview" : true,
```
Fixed in https://github.com/elastic/elasticsearch/pull/128576/commits/88b44159a053abd6d664efc56ef09e21b0b5dbc7, just hadn't run the full test in a while and missed couple

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryString.java:
```diff
@@ -105,17 +105,15 @@ public class QueryString extends FullTextFunction implements OptionalArgument {
-                lifeCycle = FunctionAppliesToLifecycle.COMING,
-                description = "Support for optional named parameters is only available in serverless, or in a future {{es}} release"
-            ) }
+            @Example(file = "qstr-function", tag = "qstr-with-options") }
```
let's follow up on that

leemthompo replied:
https://github.com/elastic/elasticsearch/pull/128576/commits/987700f2fab56b7298f49fad9f1c58ea580b2d8a

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/ToLower.java:
```diff
@@ -30,17 +28,15 @@ public class ToLower extends ChangeCase {
-            ) }
+        examples = {
+            @Example(file = "string", tag = "to_lower"),
+            @Example(file = "string", tag = "to_lower_mv", applies_to = "stack: ga 9.1.0") }
```
docs eng will handle rendering for us :)

leemthompo committed:
hide spatial functions per https://github.com/elastic/elasticsearch/pull/129839

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to

alex-spies reviewed (approved):
Thanks for the iteration @leemthompo , much easier to review now.

I have some minor remarks, but please consider this unblocked and merge after handling my comments at your own discretion.

leemthompo commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3SupportTests.java:
```diff
@@ -429,21 +415,7 @@ public static class TestClass extends Function {
-                    version = "9.0.0",
-                    description = "The ability to count the number of emojis in a string has been discontinued since 9.0.0"
-                ) }
+            appliesTo = { @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.GA, version = "9.1.0") }
```
Definitely, work for follow up

leemthompo commented on docs/reference/query-languages/esql/kibana/definition/functions/match.json:
```diff
@@ -2,7 +2,7 @@
   "comment" : "This is generated by ESQL's AbstractFunctionTestCase. Do not edit it. See ../README.md for how to regenerate it.",
   "type" : "scalar",
   "name" : "match",
-  "description" : "Use `MATCH` to perform a match query on the specified field.\nUsing `MATCH` is equivalent to using the `match` query in the Elasticsearch Query DSL.\n\nMatch can be used on fields from the text family like text and semantic_text,\nas well as other field types like keyword, boolean, dates, and numeric types.\nWhen Match is used on a semantic_text field, it will perform a semanti ...
```
I think whoever did those docs squished everything into Description which is incorrect for the normal docs, but smells like a hack to get everything into Kibana docs without updating the docs gen code to pull in DetailedDescription

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/Kql.java:
```diff
@@ -36,6 +38,8 @@ public class Kql extends FullTextFunction {
         returnType = "boolean",
         preview = true,
+        appliesTo = { @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.PREVIEW, version = "9.0.0") },
+        // @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.GA, version = "9.1.0") },
```
> But for preview, why have a version on that at all, what does it mean? The function only appeared in that version?

No it just means it was preview in 9.0, not necessarily it was added in 9.0

leemthompo commented on docs/reference/query-languages/esql/_snippets/functions/layout/match.md:
````diff
@@ -1,16 +1,10 @@
-:::
+```{applies_to}
+stack: preview 9.0.0
+serverless: preview
````
This technically doesn't change the serverless GA, because that note was only about _optional named parameters_.

The big `:::{warning}Do not use on production environments.` banner means Match is currently still in preview in the docs

leemthompo committed:
address few final comments

- added `stack: ga 9.1.0` blocks to match_phrase.md and qstr.md examples
- simplified term.md version from `preview 9.0.0` to just `preview`
- added `applies_to = "stack: ga 9.1.0"` to matchphrase and querystring java annotations
- removed version `9.0.0` from term function annotation
- deleted unused `makeCallout()` and `appendLifeCycleAndVersion()` methods

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchPhrase.java:
```diff
@@ -106,12 +107,7 @@ public class MatchPhrase extends FullTextFunction implements OptionalArgument, P
-                lifeCycle = FunctionAppliesToLifecycle.COMING,
-                description = "Support for optional named parameters is only available in serverless, or in a future {{es}} release"
-            ) }
+        examples = { @Example(file = "match-phrase-function", tag = "match-phrase-with-field") }
```
https://github.com/elastic/elasticsearch/pull/128576/commits/987700f2fab56b7298f49fad9f1c58ea580b2d8a

leemthompo commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3Support.java:
```diff
@@ -603,69 +596,42 @@ private String makeCallout(String type, String text) {
-            } else if (appliesToText.isEmpty() == false) {
-                // We have extra descriptive text, nest inside a NOTE for emphasis
-                previewText.append(makeCallout("note", appliesToText));
+        private void appendLifeCycleAndVersion(StringBuilder appliesToText, FunctionAppliesTo appliesTo) {
```
https://github.com/elastic/elasticsearch/pull/128576/commits/987700f2fab56b7298f49fad9f1c58ea580b2d8a

leemthompo committed:
Merge branch 'main' into esql_docs_real_applies_to
